### PR TITLE
When creating / editing API keys allow admins to chose whether they're visible to developers

### PR DESCRIPTION
**Pull request by siggisim**
This allows admins to choose whether or not individual API keys are visible to developers when creating / updating keys.

All existing API keys will only be visible to Admins.

Admin:
Screen Shot 2022-03-15 at 5 08 30 PM

User:
Screen Shot 2022-03-15 at 5 09 02 PM

**Version bump**: Patch

### CODE REVIEW

File: enterprise/app/api_keys/api_keys.tsx [hunk 1]
```diff
@@ -235,6 +236,15 @@ export default class ApiKeysComponent extends React.Component<ApiKeysComponentPr
     this.onChangeCapability(request, api_key.ApiKey.Capability.CACHE_WRITE_CAPABILITY, !e.target.checked, onChange);
   }
 
+  private onChangeVisibility<T extends ApiKeyFields>(
+    request: T,
+    onChange: (name: string, value: any) => any,
+    e: React.ChangeEvent<HTMLInputElement>
+  ) {
+    request.visibleToDevelopers = e.target.checked;
+    onChange("visibleToDevelopers", request.visibleToDevelopers);
```
Comment (bduffany) [inline enterprise/app/api_keys/api_keys.tsx:245]:
```suggestion
    onChange("visibleToDevelopers", e.target.checked);
```

(`onChange` winds up calling `setState` with a new request where the provided field is equal to the value passed in)

File: server/tables/tables.go [hunk 2]
```diff
@@ -305,7 +305,8 @@ type APIKey struct {
 	//
 	// NOTE: If the default is changed, a DB migration may be required to
 	// migrate old DB rows to reflect the new default.
-	Capabilities int32 `gorm:"default:1"`
+	Capabilities        int32 `gorm:"default:1"`
+	VisibleToDevelopers bool
```
Comment (bduffany) [inline server/tables/tables.go:309]: This field will be NULL by default -- dealing with nulls can sometimes be a pain, so IMO it'd be better to add the annotation `gorm:"not null"` and GORM will make this value `false` for existing rows instead

Also just double-checking, do you want this to be set to false for existing rows?

File: enterprise/server/backends/userdb/userdb.go [hunk 3]
```diff
@@ -330,7 +344,7 @@ func (d *UserDB) InsertOrUpdateGroup(ctx context.Context, g *tables.Group) (stri
 			if err := tx.Create(&newGroup).Error; err != nil {
 				return err
 			}
-			_, err = createAPIKey(tx, groupID, newAPIKeyToken(), defaultAPIKeyLabel, defaultAPIKeyCapabilities)
+			_, err = createAPIKey(tx, groupID, newAPIKeyToken(), defaultAPIKeyLabel, defaultAPIKeyCapabilities, false)
```
Comment (tylerwilliams) [inline enterprise/server/backends/userdb/userdb.go:347]: ```..., false /*visibleToDevelopers*/)``` here and below

File: enterprise/server/backends/userdb/userdb.go [hunk 4]
```diff
@@ -155,10 +156,21 @@ func (d *UserDB) GetAPIKeys(ctx context.Context, groupID string) ([]*tables.APIK
...
+	}
+
 	query := d.h.DB(ctx).Raw(`
-		SELECT api_key_id, value, label, perms, capabilities
+		SELECT api_key_id, value, label, perms, capabilities, visible_to_developers
 		FROM APIKeys
 		WHERE group_id = ?
+		`+filter+`
 		ORDER BY label ASC
 	`, groupID)
```
Comment (tylerwilliams) [inline enterprise/server/backends/userdb/userdb.go:175]: this is old code, pre-query builder, but would benefit from it. something like:

```
q := query_builder.NewQuery(`SELECT api_key_id, value, label, perms, capabilities, visible_to_developers FROM APIKeys`)
q.AddWhereClause("group_id = ?", groupID)
q.SetOrderBy("label", true /*ascending*/)
if err := authutil.AuthorizeGroupRole(u, groupID, role.Admin); err == nil {
		q.AddWhereClause("visible_to_developers = ?", true)
}
query := d.h.DB(ctx).Raw(q.Build())
```

**Review by tylerwilliams — APPROVED** [review]
LGTM w/ a few minor nits

this is sweet

File: server/tables/tables.go (hunk 2 above)
Comment (siggisim) [reply server/tables/tables.go]: Yeah, I definitely want all API keys that are visible to developers to be manually marked as such.

I've added `gorm:"not null;default:false"` to make the default value false.

**Pull request merged**